keras/engine/training.py

```diff
@@ -3258,6 +3258,24 @@ def concat(tensors, axis=0):
   return tf.concat(tensors, axis=axis)
 
 
+def potentially_variable_concat(tensors):
+  """Concats `tensor`s along `axis`. When non-batch dimensions are variable, a RaggedTensor is returned instead."""
+  if isinstance(tensors[0], tf.SparseTensor):
+    return tf.sparse.concat(axis=0, sp_inputs=tensors)
+  non_batch_shapes = tf.stack([tf.shape(tensor)[1:] for tensor in tensors])
+  constant_dims = tf.math.reduce_all(non_batch_shapes == non_batch_shapes[:1], axis=0)
+  if tf.math.reduce_all(constant_dims).numpy().item():  # All non-batch dimensions are constant
+    return tf.concat(tensors, axis=0)
+  # First, identify constant inner dimensions by finding the rightmost dimension that is not constant
```

Comment:
Because of the `reduce_all`, I believe this conditional is only `True` if all dimensions are constant - the tensor will only have a single element at the point when we call `item()`. Let me know if I overlooked something, though!

Reply:
Note that `non_batch_shapes` is a 2D tensor of shape `(num_batches, num_non_batch_dims)` - the first `reduce_all` flattens that to `(num_non_batch_dims,)` and the second flattens it to a single element.